src/circuit/qc_keyagg.rs

```diff
@@ -1,26 +1,71 @@
 //! Circuit implementation of stake key aggregation for quorum certificates verification.
-use ark_ec::{
-    short_weierstrass::{Projective, SWCurveConfig},
-    CurveGroup,
-};
+
 use ark_ff::PrimeField;
-use ark_std::{format, vec, Zero};
-use jf_primitives::{circuit::rescue::RescueNativeGadget, rescue::RescueParameter};
+use ark_std::{format, vec, vec::Vec};
+use jf_primitives::{
+    circuit::rescue::RescueNativeGadget,
+    rescue::{sponge::RescueCRHF, RescueParameter},
+};
 use jf_relation::{
     errors::CircuitError,
     gadgets::{
-        ecc::{non_native::EmulatedPointVariable, Point, SWToTEConParam},
-        EmulationConfig,
+        ecc::{
+            emulated::{EmulatedSWPointVariable, EmulatedTEPointVariable, SWPoint},
+            TEPoint,
+        },
+        EmulationConfig, SerializableEmulatedStruct,
     },
     BoolVar, Circuit, PlonkCircuit, Variable,
 };
 
-#[derive(Debug, Clone)]
-/// Stake public key variable
-/// Wrap EmulatedPointVariable because we need to simulate curve operations over the same curve's scalar field
-pub struct VerKeyVar<E: PrimeField>(pub EmulatedPointVariable<E>);
+pub fn compute_stake_table_hash<F: RescueParameter, T: SerializableEmulatedStruct<F>>(
+    stake_amts: &[F],
+    vk_points: &[T],
+) -> F {
+    let mut input_vec = vec![];
+    for (&amt, point) in stake_amts.iter().zip(vk_points.iter()) {
+        input_vec.extend(point.serialize_to_native_elements());
+        input_vec.push(amt);
+    }
+    RescueCRHF::sponge_with_bit_padding(&input_vec[..], 1)[0]
+}
+
+pub trait VerKeyVar<E>: Sized + Clone {
+    type KeyType: Default;
+
+    fn native_vars(&self) -> Vec<Variable>;
+
+    fn aggregate_with_selectors<F>(
+        circuit: &mut PlonkCircuit<F>,
+        keys: &[Self],
+        selectors: &[BoolVar],
+        coef: E,
+    ) -> Result<Self, CircuitError>
+    where
+        F: PrimeField,
+        E: EmulationConfig<F>;
+
+    fn is_equal<F>(
+        circuit: &mut PlonkCircuit<F>,
+        p0: &Self,
+        p1: &Self,
+    ) -> Result<BoolVar, CircuitError>
+    where
+        F: PrimeField,
+        E: EmulationConfig<F>;
+
+    fn enforce_equal<F>(
+        circuit: &mut PlonkCircuit<F>,
+        p0: &Self,
+        p1: &Self,
+    ) -> Result<(), CircuitError>
+    where
+        F: PrimeField,
+        E: EmulationConfig<F>;
+}
 
 /// Plonk circuit gadget for stake key aggregation for quorum certificates.
+/// Assuming that the underlying key composes of a short Weierstrass curve point.
```

Comment:
Do we want to have this assumption for a general trait?

Reply:
Removed in [8697221](https://github.com/EspressoSystems/hotshot-primitives/pull/74/commits/869722169145e40b275b9d16e50284a3868e2a60)
 No assumption now.